modules/4337/src/utils/execution.ts

```diff
@@ -28,44 +44,85 @@ export interface SafeTransaction extends MetaTransaction {
   const typedDataHash = ethers.getBytes(hash)
+  const signerAddress = await signer.getAddress()
   return {
-    signer: await signer.getAddress(),
-    data: await signer.signMessage(typedDataHash),
+    signer: signerAddress,
+    data: (await signer.signMessage(typedDataHash)).replace(/1b$/, '1f').replace(/1c$/, '20'),
   }
 }
 
-const sortSignatures = (signatures: SafeSignature[]) => {
-  signatures.sort((left, right) => left.signer.toLowerCase().localeCompare(right.signer.toLowerCase()))
+export const buildContractSignature = (signerAddress: string, signature: string): SafeSignature => {
+  return {
+    signer: signerAddress,
+    data: signature,
+    dynamic: true,
+  }
 }
 
 export const buildSignatureBytes = (signatures: SafeSignature[]): string => {
-  sortSignatures(signatures)
-  return ethers.concat(signatures.map((signature) => signature.data))
-}
-
-export const buildContractSignatureBytes = (signatures: SafeSignature[]): string => {
-  sortSignatures(signatures)
-  const start = 65 * signatures.length
-  const { segments } = signatures.reduce(
-    ({ segments, offset }, { signer, data }) => {
-      return {
-        segments: [...segments, ethers.solidityPacked(['uint256', 'uint256', 'uint8'], [signer, start + offset, 0])],
-        offset: offset + 32 + ethers.dataLength(data),
-      }
-    },
-    { segments: [] as string[], offset: 0 },
-  )
-  return ethers.concat([
-    ...segments,
-    ...signatures.map(({ data }) => ethers.solidityPacked(['uint256', 'bytes'], [ethers.dataLength(data), data])),
-  ])
+  const SIGNATURE_LENGTH_BYTES = 65
+  signatures.sort((left, right) => left.signer.toLowerCase().localeCompare(right.signer.toLowerCase()))
+
+  let signatureBytes = '0x'
+  let dynamicBytes = ''
+  for (const sig of signatures) {
+    if (sig.dynamic) {
+      /* 
+              A contract signature has a static part of 65 bytes and the dynamic part that needs to be appended 
+              at the end of signature bytes.
+              The signature format is
+              Signature type == 0
+              Constant part: 65 bytes
+              {32-bytes signature verifier}{32-bytes dynamic data position}{1-byte signature type}
+              Dynamic part (solidity bytes): 32 bytes + signature data length
+              {32-bytes signature length}{bytes signature data}
+          */
+      const dynamicPartPosition = (signatures.length * SIGNATURE_LENGTH_BYTES + dynamicBytes.length / 2).toString(16).padStart(64, '0')
+      const dynamicPartLength = (sig.data.slice(2).length / 2).toString(16).padStart(64, '0')
+      const staticSignature = `${sig.signer.slice(2).padStart(64, '0')}${dynamicPartPosition}00`
+      const dynamicPartWithLength = `${dynamicPartLength}${sig.data.slice(2)}`
+
+      signatureBytes += staticSignature
+      dynamicBytes += dynamicPartWithLength
+    } else {
+      signatureBytes += sig.data.slice(2)
+    }
+  }
+
+  return signatureBytes + dynamicBytes
```

Comment:
This whole file is a copy-paste from the core safe contracts repo, which is suboptimal. The issue is that the Safe Contracts repo uses ethers v5, and we use ethers v6, and they're incompatible. Perhaps we should consider releasing a package with 1.4.1 contracts and ethers v6.

Reply:
We can add a tech debt item for the next sprint. We want to start working on Safe account contracts in the near future, so the work would be strategic as well.

Reply:
> We can add a tech debt item for the next sprint. We want to start working on Safe account contracts in the near future, so the work would be strategic as well.

This would require some planning because we already upgraded the repo with ethers v6 in the upcoming 1.5.0. The question we'd need to answer is if we need to release an incremental version of 1.4.1 with this upgrade or just wait for 1.5.0.